File: crates/p2a-desktop/ui/src/lib/api/llm.ts
// LLM API wrappers for Tauri commands

import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import type {
	Conversation,
	LlmMessage,
	ProviderConfig,
	SendMessageResponse,
	StreamChunk
} from '$lib/types/llm';

/**
 * Send a message to the LLM and get a complete response.
 * @param message The user's message
 * @param conversationId Optional existing conversation ID
 */
export async function sendMessage(
	message: string,
	conversationId?: string
): Promise<SendMessageResponse> {
	return invoke<SendMessageResponse>('send_message', {
		message,
		conversationId
	});
}

/**
 * Send a message with streaming response.
 * Use listenToStream to receive stream events.
 * @param message The user's message
 * @param conversationId Optional existing conversation ID
 */
export async function sendMessageStream(
	message: string,
	conversationId?: string
): Promise<SendMessageResponse> {
	return invoke<SendMessageResponse>('send_message_stream', {
		message,
		conversationId
	});
}

/**
 * Listen to LLM stream events.
 * @param callback Function to handle stream chunks
 * @returns Unlisten function to stop listening
 */
export async function listenToStream(
	callback: (chunk: StreamChunk) => void
): Promise<UnlistenFn> {
	return listen<StreamChunk>('llm-stream', (event) => {
		callback(event.payload);
	});
}

/**
 * List all conversations.
 */
export async function listConversations(): Promise<Conversation[]> {
	return invoke<Conversation[]>('list_conversations');
}

/**
 * Get messages for a conversation.
 * @param conversationId The conversation ID
 */
export async function getConversation(conversationId: string): Promise<LlmMessage[]> {
	return invoke<LlmMessage[]>('get_conversation', { conversationId });
}

/**
 * Delete a conversation.
 * @param conversationId The conversation ID
 */
export async function deleteConversation(conversationId: string): Promise<void> {
	return invoke('delete_conversation', { conversationId });
}

/**
 * Get current LLM settings.
 */
export async function getLlmSettings(): Promise<ProviderConfig> {
	return invoke<ProviderConfig>('get_llm_settings');
}

/**
 * Update LLM settings.
 * @param config The new configuration
 */
export async function updateLlmSettings(config: ProviderConfig): Promise<void> {
	return invoke('update_llm_settings', { config });
}

/**
 * Check if the current provider is available.
 */
export async function checkProvider(): Promise<boolean> {
	return invoke<boolean>('check_provider');
}

/**
 * List available models for the current provider.
 */
export async function listProviderModels(): Promise<string[]> {
	return invoke<string[]>('list_provider_models');
}

/**
 * Get list of supported provider types.
 */
export async function listProviderTypes(): Promise<string[]> {
	return invoke<string[]>('list_provider_types');
}

/**
 * Rename a conversation.
 * @param conversationId The conversation ID
 * @param newTitle The new title
 */
export async function renameConversation(conversationId: string, newTitle: string): Promise<void> {
	return invoke('rename_conversation', { conversationId, newTitle });
}

/** Exported conversation data */
export interface ExportedConversation {
	id: string;
	title: string;
	created_at: string;
	updated_at: string;
	provider: string;
	model: string;
	messages: LlmMessage[];
}

/**
 * Export a conversation with all metadata and messages.
 * @param conversationId The conversation ID
 */
export async function exportConversation(conversationId: string): Promise<ExportedConversation> {
	return invoke<ExportedConversation>('export_conversation', { conversationId });
}

File: crates/p2a-desktop/src/commands/llm.rs
//! LLM-related Tauri commands.

use crate::llm::{Conversation, Message, ProviderConfig, ProviderType, StreamChunk};
use crate::AppState;
use tauri::{Emitter, State};

/// Response from sending a message.
#[derive(serde::Serialize)]
pub struct SendMessageResponse {
    pub conversation_id: String,
    pub message: Message,
}

/// Send a message to the LLM and get a response.
#[tauri::command]
pub async fn send_message(
    state: State<'_, AppState>,
    conversation_id: Option<String>,
    message: String,
) -> Result<SendMessageResponse, String> {
    let llm_service = state.llm_service();

    // Ensure MCP server is running for tool execution
    let mcp_client = state.mcp_client();
    if !mcp_client.is_running() {
        mcp_client.spawn().await.map_err(|e| e.to_string())?;
    }

    let (conv_id, response) = llm_service
        .send_message(conversation_id.as_deref(), &message)
        .await
        .map_err(|e| e.to_string())?;

    Ok(SendMessageResponse {
        conversation_id: conv_id,
        message: response,
    })
}

/// Send a message with streaming response.
/// Emits "llm-stream" events to the frontend.
#[tauri::command]
pub async fn send_message_stream(
    state: State<'_, AppState>,
    app: tauri::AppHandle,
    conversation_id: Option<String>,
    message: String,
) -> Result<SendMessageResponse, String> {
    let llm_service = state.llm_service();

    // Ensure MCP server is running
    let mcp_client = state.mcp_client();
    if !mcp_client.is_running() {
        mcp_client.spawn().await.map_err(|e| e.to_string())?;
    }

    // Create callback that emits events to frontend
    let app_handle = app.clone();
    let callback = Box::new(move |chunk: StreamChunk| {
        let _ = app_handle.emit("llm-stream", &chunk);
    });

    let (conv_id, response) = llm_service
        .send_message_stream(conversation_id.as_deref(), &message, callback)
        .await
        .map_err(|e| e.to_string())?;

    Ok(SendMessageResponse {
        conversation_id: conv_id,
        message: response,
    })
}

/// List all conversations.
#[tauri::command]
pub async fn list_conversations(
    state: State<'_, AppState>,
) -> Result<Vec<Conversation>, String> {
    let llm_service = state.llm_service();
    llm_service
        .list_conversations()
        .map_err(|e| e.to_string())
}

/// Get messages for a conversation.
#[tauri::command]
pub async fn get_conversation(
    state: State<'_, AppState>,
    conversation_id: String,
) -> Result<Vec<Message>, String> {
    let llm_service = state.llm_service();
    llm_service
        .get_conversation_messages(&conversation_id)
        .map_err(|e| e.to_string())
}

/// Delete a conversation.
#[tauri::command]
pub async fn delete_conversation(
    state: State<'_, AppState>,
    conversation_id: String,
) -> Result<(), String> {
    let llm_service = state.llm_service();
    llm_service
        .delete_conversation(&conversation_id)
        .map_err(|e| e.to_string())
}

/// Get current LLM settings.
#[tauri::command]
pub async fn get_llm_settings(state: State<'_, AppState>) -> Result<ProviderConfig, String> {
    let llm_service = state.llm_service();
    Ok(llm_service.get_config().await)
}

/// Update LLM settings.
#[tauri::command]
pub async fn update_llm_settings(
    state: State<'_, AppState>,
    config: ProviderConfig,
) -> Result<(), String> {
    let llm_service = state.llm_service();
    llm_service
        .set_config(config)
        .await
        .map_err(|e| e.to_string())
}

/// Check if current provider is available.
#[tauri::command]
pub async fn check_provider(state: State<'_, AppState>) -> Result<bool, String> {
    let llm_service = state.llm_service();
    llm_service
        .is_provider_available()
        .await
        .map_err(|e| e.to_string())
}

/// List available models for current provider.
#[tauri::command]
pub async fn list_provider_models(state: State<'_, AppState>) -> Result<Vec<String>, String> {
    let llm_service = state.llm_service();
    llm_service
        .list_models()
        .await
        .map_err(|e| e.to_string())
}

/// Get list of supported provider types.
#[tauri::command]
pub fn list_provider_types() -> Vec<String> {
    vec![
        ProviderType::Ollama.to_string(),
        ProviderType::Anthropic.to_string(),
        ProviderType::OpenAI.to_string(),
    ]
}

/// Rename a conversation.
#[tauri::command]
pub async fn rename_conversation(
    state: State<'_, AppState>,
    conversation_id: String,
    new_title: String,
) -> Result<(), String> {
    let llm_service = state.llm_service();
    llm_service
        .rename_conversation(&conversation_id, &new_title)
        .map_err(|e| e.to_string())
}

/// Export conversation data structure.
#[derive(serde::Serialize)]
pub struct ExportedConversation {
    pub id: String,
    pub title: String,
    pub created_at: String,
    pub updated_at: String,
    pub provider: String,
    pub model: String,
    pub messages: Vec<Message>,
}

/// Export a conversation.
#[tauri::command]
pub async fn export_conversation(
    state: State<'_, AppState>,
    conversation_id: String,
) -> Result<ExportedConversation, String> {
    let llm_service = state.llm_service();

    // Get conversation metadata
    let conversation = llm_service
        .get_conversation(&conversation_id)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Conversation not found".to_string())?;

    // Get messages
    let messages = llm_service
        .get_conversation_messages(&conversation_id)
        .map_err(|e| e.to_string())?;

    Ok(ExportedConversation {
        id: conversation.id,
        title: conversation.title,
        created_at: conversation.created_at.to_rfc3339(),
        updated_at: conversation.updated_at.to_rfc3339(),
        provider: conversation.provider.to_string(),
        model: conversation.model,
        messages,
    })
}
